Implement fuse-prototype including context-mocking and README.md (#3657)
ElektraInitiative/libelektra: scripts/fuse/README.MD

# elektra-fuse

Mount  [elektra](https://github.com/ElektraInitiative/libelektra) as a [fuse(py)](https://github.com/fusepy/fusepy) filesystem.

This tool enables the inspection and modification of the Elektra-hierarchy, in the form of a classical filesystem, as it is observed by a given running process.
In the simplest case, an Elektra key appears as a file of the same name, where its contents pose the keys value.

This enables users to interact with Elektra
- using standard unix tooling, making Elektra more accessible
- with standard system/library procedures
- without the need to build with libelektra
- with additional features: e.g. the ability to effortlessly write binary keys

## Quickstart

### Docker

Ensure docker is installed and run:
```sh

> ./run_new_container.sh

Reviewer: @kodebach were you able to reproduce?

> ./run_new_container.sh

Reviewer 2: I do get this message, but afterwards the container still builds and starts up. The script just doesn't account for the fact that there might not be an existing container. Some `if`s could fix that, but a simple `> /dev/null` might work too.

> ./run_new_container.sh

Author: Resolved in 2d8bd32adaa0eb746f3787ca3db02d80bdf5c054

```
This (re)creates the container if necessary and starts an interactive session with keys for testing already created (see `docker/create_keys.sh`).
The filesystem is mounted below `~/mount`.
For debugging purposes, a logfile is written to `~/nohup.out` inside the container. 

### Native

Required (debian) packages:
- libfuse-dev
- fuse

> - libelektra5-all (not in official repositories, follow [these](https://www.libelektra.org/installation/installation) instructions)

Author: Dependencies are listed, nothing needed for building as nothing is built. Minimum Elektra version mentioned as of f0cd885d86f02425bcf208ddad30bafc02ce3aa2

Required (python3) packages:
- fusepy psutil

Mount the filesystem below the (already existing) directory `mountpoint` (as root-user) with
```sh
sudo ./elektra_fuse/elektra_fuse.py `mountpoint`
```

## Filesystem Structure

Directly below the mount point, for every process running on the system (except those whose current working directory lies below said mountipoint, e.g. the shell exporing the hierarchy), a directory named after the respective process identifier is exposed.

For convinience, each such directory possesses numerous extened filesystem attributes that identify the process, user, etc., that can be listed by:

```sh
xattr -l mountpoint/<pid>
```

Below any such pid-directory, the elektra key database is mounted bidirectionally using directories that correspond to the different namespaces:

- `user:`
- `system:`
- `dir:`
- `spec:`

Exceptions (at the moment, these namespaces behave differently)

> - `default:`

Author: Readme updated (f0cd885d86f02425bcf208ddad30bafc02ce3aa2)

> - `proc:`

Author: Easily implementable in future version, mentioned in f0cd885d86f02425bcf208ddad30bafc02ce3aa2

Furthermore, there is `cascading:`, which correspons to the `/` Elektra-namespace.

> This is useful for inspection purposes, but beware:

Author: Elaborated in next thesis draft

> This is useful for inspection purposes, but beware:

Author: Cascading writes in the FUSE implementation are disabled since e42cfd8a1eec378550556d4242a0a5ee2ac1a095.

See also #3794.

Write operations in this namespace produce unexpected effects (and create operations might fail), as for the same "address" a write operation can be resolved to a different real key then the corresponding read operation.

> These first two layers, (e.g `mountpoint/12/system:`) are read only. Deeper layers support writing: any change is directly updated in the backing Elektra-Key-Database.

Reviewer: Btw. it is good if you also push a commit fixing the problem when you resolve conversations.

### Process Context Mocking

All Elektra operations are performed "from the perspective" of the respective process.
This is done by only calling libelektra/kdb from special "mock"-processes that mimic all of the relevant attributes Elektra uses to resolve its keys.

The mocked attributes are:
- the user and group (for the `user:` namespace)
- working directory (for the `dir:` namespace)
- environment variables (for the `proc:` namespace)

> - in the future: the process arguments

Author: Could you further elaborate? (What exactly is already in the proc: namespace?)

> - in the future: the process arguments

Reviewer: command-line arguments and environment variables, see also https://github.com/ElektraInitiative/libelektra/pull/3657#discussion_r606660739

> - in the future: the process arguments

Author: See https://github.com/ElektraInitiative/libelektra/pull/3657#discussion_r615415441.

> - in the future: the process arguments

Reviewer: This line in the README.md still needs to be updated, though.

```
|-- 41
|   |-- cascading:
|   |   |-- dir_and_file_at_once
|   |   |   `-- leaf
|   |   |-- dirkey
|   |   |-- elektra
|   |   |   |-- modules
|   |   |   `-- version
|   |   |-- info
|   |   `-- person
|   |       `-- name
|   `-- user:
|       |-- dir_and_file_at_once
|       |   |-- @elektra.value
|       |   `-- leaf
|       |-- info
|       `-- person
|           `-- name

```
as obtained with
```sh
tree -L 4 mountpoint
```

## Assumptions and special behaviour

Elektra cannot map perfectly to the structure of a classical filesystem as described [here](https://www.libelektra.org/docgettingstarted/big-picture).
Whenever deviations are necessary, the goal is to maximize usability and adhere to the expectations of common Unix tools.

### Intermediate directories

In Elektra, there is no need to have a `path` of continous keys between ancestors.
E.g. the existence of `user:/a` and `user:/a/b/c` does not imply the existence of `user:/a/b`.
Such intermediate keys are mapped to "virtual" directories, s.t. navigation in an interactive manner is possible and the directory tree is continous.

> A side effect is that creating/deleting a single key can create/delete multiple "virtual" directories at once.

Author: `touch` and `mkdir` do behave differently in the implementation (mkdir creates the meta-key "fuse/directory")

> A side effect is that creating/deleting a single key can create/delete multiple "virtual" directories at once.

Author: I added an example as you proposed in 63e59cfd461611eca3f219c832beb5ae97c5b9c9.

Another is that copies created with `cp -r` will not have these vacuous intermediaries but instead empty keys.

> 

Author: See cd892f3ce0a60971894e8b959af1a736eccdfd8c

> Another is that copies created with `cp -r` will not have these vacuous intermediaries but instead empty keys.

Author: See cd892f3ce0a60971894e8b959af1a736eccdfd8c

### Keys that act both as file and directory

In Elektra, there are only keys and no distinction between directories and files are made. To enable a bidirectional mapping into a classical file system, the following mechanism is employed:

If a key (`A`) exists, and has some child key (`B`), `A` is treated as a directory containing `B` and a special file `@elektra.value`. This file is used to access the value of `A`.

Some side effects arise:

> E.g. recursively deleting directories with `rm -r` containing `@elektra.value` files raises (task uncritical) errors when attempting to delete a parent directory which was already deleted with the removal of an `@elektra.value` file directly below.

Reviewer: ```suggestion
E.g. recursively deleting directories with `rm -r` containing `®elektra.value` files raises (task uncritical) errors when attempting to delete a parent directory which was already deleted with the removal of an `®elektra.value` file directly below.
```

see #3599

Reviewer: >Some (older) file systems may have trouble with the ® character (or any other non-ASCII characters).

As we implement a file system here, this should not be a problem. Terminals or language settings not supporting the ® character, however, can be a problem. In specific, the fuse Docker image we provide doesn't have locales installed, so ® does not work (at least when I tried the last time).

In any case, we should follow our own standard, so either, we

1. use ®elektra
2. also reserve ®elektra.value (or even ®elektra.*)

Reviewer 2: I would suggest we use the `®elektra` folder if possible. From a general point of view it makes more sense for Elektra to only reserve `®elektra` as a key name part. Since the KDB is hierarchical anyway there is no need to reserve all key name parts that start with X.

Reviewer: >    reduces usability (one file versus one directory with always one file in it)

Yes, it is pressing 2x tab instead of one. The question is if there is a possibility that more files get added in the future, e.g. for exporting functionality?

> Furthermore, again because of usability considerations, I'm not happy with the use of ® instead of @ due to the mentioned encoding problems. Imho, any such naming clash is extremely unlikely in practical use cases, whereas the pain of obfuscated encodings is real.

I see your point. Is there also a possibility of obfuscated encodings except if the locales package is not installed at all? We chose ® exactly because it also has somewhat reasonable encoding if the wrong locale is chosen (even in ASCII).

@kodebach there is still a "TODO: link to detailed ®elektra document" in doc/KEYNAMES.md. Would be better to have this document then long discussions here in this sub-thread never to be found again. We can continue the discussion in the PR.

I am open to `@elektra` but then `@elektra` and not `®elektra` should be the reserved key name.

Btw. I didn't see the initial response https://github.com/ElektraInitiative/libelektra/pull/3657#discussion_r604835066 of @dev2718 when I wrote https://github.com/ElektraInitiative/libelektra/pull/3657#issuecomment-814239192

Author: As of f13604c3f71c210cbc9dbc115069d96ec50ba9ad `®elektra` is standard, but configurable via command line option.

### "Leaf"-Keys that act as as directories

Keys at the bottom of the hierarchy are normally treaded as files. However in certain cases, like explicitly creating directories using `mkdir`, this would result in files beeing displayed, making it impossible to expand the hierarchy using standard tools.

> Therefore, when a directory is created, the special meta-key `fuse-directory` is set to override this behaviour.

Reviewer: use / as separator within (meta)keys.

```suggestion
Therefore, when a directory is created, the special meta-key `fuse/directory` is set to override this behaviour.
```

### Filesystem attributes

> The filesystem attributes are sourced from the file obtained by `kdb file`.

Reviewer: Is it also possible to receive the file name itself? see #3599

What is the meaning of "number of links or directories", it seems to differ for the real file and the FUSE file?

Author: Since 66acb118f758035459212e3e730bc4e001d430b9 the functionality of `kdb file` is implemented natively.
Or do you propose the ability to query the file name for example as an xattr?

Reviewer: Is this maybe also a candidate for a file below `®elektra`?

Author: See in https://github.com/ElektraInitiative/libelektra/pull/3657#discussion_r615417258.

For some keys, no such file is returned, or the returned file does not exist in actuality. In these cases, fake attributes are used to preserve usability.

> To increase compatibility, currently the timestamps from these files are not used, as this prompts software like `vim` to produce warnings/errors.

Reviewer: Which time stamps? The modification date seems to be present.

## Meta-Keys

Meta-keys are mapped bidirectionally to extended file attributes (xattrs).

> To enable interoperability with other tools, the `meta:/` prefix is optional when reading/writing accessing meta keys. The listing of attributes however will show them prefix-less.

Author: Resolved in a4cc4e0cf020690b0601f486f4394e94d84b3c58

Furthermore, no binary values for these attributes are allowed. (As libelektra does not accept those).

> See `xattr(1)` on how to read/write/list these attributes.

Reviewer: The man page is not present in Docker and on my system (7) opens by default which is full of descriptions of file-system-specific behavior. A small example how to get+set will not harm, e.g.:

```
xattr -w abc def person/name
kdb meta-get system:/person/name abc
```

## Binary-Keys

If a value is to be written that cannot be decoded using the systems default encoding, it is treated as binary, and the appropriate `meta:/binary` meta-key is set.

## Known issues

> - Moving an ``@elektra.value``-file corrupts the filesytsem/key database (See this [issue](https://github.com/ElektraInitiative/libelektra/issues/3648)). As editors like `vim` may move a file during editing, such editors cannot reliebly be used on these pseudo-files.

Reviewer: ```suggestion
- Moving an ``@elektra.value``-file corrupts the filesystem/key database (See this [issue](https://issues.libelektra.org/3648)). As editors like `vim` may move a file during editing, such editors cannot reliably be used on these pseudo-files.
```

> - Only functional on UNIX-like systems

Reviewer: Isn't the limitation simply that it will only work on systems that support FUSE?

```suggestion
- Only functional on Unix-like systems
```

> - chmod/chown is currently not implemented (and how it should be is not obvious). A call does not signal a "not supported error" to enable compatibility with common tools like `cp`.

Reviewer: ```suggestion
- chmod/chown is not implemented and does not signal a "not supported error" to enable compatibility with common tools like `cp`.
```

> - Access related errorcodes not yet implemented (e.g. normal user tries to write to system:/) (defaulting to read only fs - error)

Reviewer: I see reporting Elektra's errors to the user within the scope of the thesis.

> - Access related errorcodes not yet implemented (e.g. normal user tries to write to system:/) (defaulting to read only fs - error)

Author: As far as I know FUSE does not support custom error messages, we only have the standard UNIX-error codes available.

> - Access related errorcodes not yet implemented (e.g. normal user tries to write to system:/) (defaulting to read only fs - error)

Reviewer: Most file systems simply log what they wanted to say if they cannot via errno, e.g. they report file corruption that way. KDB reporting errors is very similar as it is also "the layer below". You could simply use [syslog](https://docs.python.org/3/library/syslog.html).

Special files would be another option to report errors (like lost+found), but that would conflict with your stateless approach.

> - Access related errorcodes not yet implemented (e.g. normal user tries to write to system:/) (defaulting to read only fs - error)

Author: Just to avoid any misunderstandings:

I am supposed to catch (unexpected, e.g. binary writes to meta-keys) Elektra exceptions and log them via syslog in my implementation?

> - Access related errorcodes not yet implemented (e.g. normal user tries to write to system:/) (defaulting to read only fs - error)

Reviewer: No, things like binary writes to meta-keys can be done via normal errno codes maybe with ENOTSUP.

The logger would be used when kdbOpen/kdbGet/kdbSet return errors.

> - Access related errorcodes not yet implemented (e.g. normal user tries to write to system:/) (defaulting to read only fs - error)

Author: As of 83d67af36959a3181e9641ccd1cf299ce10961f5 there is configurable logging (via command line options:(stdout/syslog/none)).  Exceptions thrown by Elektra are also logged.

Required (debian) packages:
- libfuse-dev
- fuse
- libelektra5-all (not in official repositories, follow [these](https://www.libelektra.org/installation/installation) instructions)

Required (python3) packages:
- fusepy psutil

Mount the filesystem below the (already existing) directory `<mount point>` (as root-user) with
```sh
sudo ./elektra_fuse/elektra_fuse.py <mount point>
```

## Filesystem Structure

Directly below the mount point, for every process running on the system (except those whose current working directory lies below said mountipoint, e.g. the shell exporing the hierarchy), a directory named after the respective process identifier is exposed.

For convinience, each such directory possesses numerous extened filesystem attributes that identify the process, user, etc., that can be listed by:

```sh
xattr -l <mount point>/<pid>
```

Below any such pid-directory, the elektra key database is mounted bidirectionally (i.e. read-write capable) using directories that correspond to the different [namespaces](https://www.libelektra.org/tutorials/namespaces):

- `user:`
- `system:`
- `dir:`
- `spec:`

Exceptions:
- `default:` (only exists for cascading lookups)

> - `proc:` (to be implemented in future version (read-only))

Reviewer: Docu not updated?

> Furthermore, there is `cascading:`, which correspons to the `/` Elektra-namespace.

Reviewer: Why is it not in the list above?
